### ISSUE TITLE
Fix drop type checking for nested `Arc`/`Mutex` objects

### ISSUE THREAD

**Issue by avanhatt** [model-checking/kani#426]
The two `SizeAndAlignOfDst` tests fail with:

```rust
"func":"Expr"{
   "value":"Symbol"{
      "identifier":"_RINvNtCsgWci0eQkB8o_4core3ptr13drop_in_placeNtNtNtCs1HAdiQHUxxm_3std10sys_common5mutex12MovableMutexECsb7rQPrKk64Y_4main"
   },
   "typ":"Code"{
      "parameters":[
         "Parameter"{
            "typ":"Pointer"{
               "typ":"StructTag(""tag-std::sys_common::mutex::MovableMutex"")"
            },
            "identifier":"None",
            "base_name":"None"
         }
      ],
      "return_type":"StructTag(""tag-Unit"")"
   },
   "location":"None"
}"args":[
   "Expr"{
      "value":"Symbol"{
         "identifier":"_RINvNtCsgWci0eQkB8o_4core3ptr13drop_in_placeINtNtNtCs1HAdiQHUxxm_3std4sync5mutex5MutexDNtCsb7rQPrKk64Y_4main10SubscriberEL_EEB1p_::1::var_1"
      },
      "typ":"StructTag(""tag-dyn Subscriber"")",
      "location":"None"
   }
]"', compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/expr.rs:532:9"

stack backtrace:
   0: rust_begin_unwind
             at ./library/std/src/panicking.rs:517:5
   1: std::panicking::begin_panic_fmt
             at ./library/std/src/panicking.rs:460:5
   2: rustc_codegen_llvm::gotoc::cbmc::goto_program::expr::Expr::call
             at ./compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/expr.rs:532:9
   3: rustc_codegen_llvm::gotoc::statement::<impl rustc_codegen_llvm::gotoc::metadata::GotocCtx>::codegen_drop
             at ./compiler/rustc_codegen_llvm/src/gotoc/statement.rs:148:29
   4: rustc_codegen_llvm::gotoc::statement::<impl rustc_codegen_llvm::gotoc::metadata::GotocCtx>::codegen_terminator
             at ./compiler/rustc_codegen_llvm/src/gotoc/statement.rs:58:66
   5: rustc_codegen_llvm::gotoc::block::<impl rustc_codegen_llvm::gotoc::metadata::GotocCtx>::codegen_block
             at ./compiler/rustc_codegen_llvm/src/gotoc/block.rs:18:29
   6: rustc_codegen_llvm::gotoc::function::<impl rustc_codegen_llvm::gotoc::metadata::GotocCtx>::codegen_function::{{closure}}
             at ./compiler/rustc_codegen_llvm/src/gotoc/function.rs:95:71
   7: core::iter::traits::iterator::Iterator::for_each::call::{{closure}}
             at ./library/core/src/iter/traits/iterator.rs:730:29
   8: core::iter::adapters::map::map_fold::{{closure}}
             at ./library/core/src/iter/adapters/map.rs:84:21
   9: <core::iter::adapters::enumerate::Enumerate<I> as core::iter::traits::iterator::Iterator>::fold::enumerate::{{closure}}
             at ./library/core/src/iter/adapters/enumerate.rs:106:27
  10: core::iter::traits::iterator::Iterator::fold
             at ./library/core/src/iter/traits/iterator.rs:2170:21
  11: <core::iter::adapters::enumerate::Enumerate<I> as core::iter::traits::iterator::Iterator>::fold
             at ./library/core/src/iter/adapters/enumerate.rs:112:9
  12: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
             at ./library/core/src/iter/adapters/map.rs:124:9
  13: core::iter::traits::iterator::Iterator::for_each
             at ./library/core/src/iter/traits/iterator.rs:733:9
  14: rustc_codegen_llvm::gotoc::function::<impl rustc_codegen_llvm::gotoc::metadata::GotocCtx>::codegen_function
             at ./compiler/rustc_codegen_llvm/src/gotoc/function.rs:95:13
  15: <rustc_codegen_llvm::gotoc::backend::GotocCodegenBackend as rustc_codegen_ssa::traits::backend::CodegenBackend>::codegen_crate::{{closure}}
             at ./compiler/rustc_codegen_llvm/src/gotoc/backend.rs:102:35
  16: rustc_codegen_llvm::gotoc::debug::<impl rustc_codegen_llvm::gotoc::metadata::GotocCtx>::call_with_panic_debug_info::{{closure}}
             at ./compiler/rustc_codegen_llvm/src/gotoc/debug.rs:72:13
  17: std::thread::local::LocalKey<T>::try_with
             at ./library/std/src/thread/local.rs:399:16
  18: std::thread::local::LocalKey<T>::with
             at ./library/std/src/thread/local.rs:375:9
  19: rustc_codegen_llvm::gotoc::debug::<impl rustc_codegen_llvm::gotoc::metadata::GotocCtx>::call_with_panic_debug_info
             at ./compiler/rustc_codegen_llvm/src/gotoc/debug.rs:70:9
  20: <rustc_codegen_llvm::gotoc::backend::GotocCodegenBackend as rustc_codegen_ssa::traits::backend::CodegenBackend>::codegen_crate
             at ./compiler/rustc_codegen_llvm/src/gotoc/backend.rs:101:25
  21: rustc_interface::passes::start_codegen::{{closure}}
             at ./compiler/rustc_interface/src/passes.rs:1056:9
  22: rustc_data_structures::profiling::VerboseTimingGuard::run
             at ./compiler/rustc_data_structures/src/profiling.rs:611:9
  23: rustc_session::utils::<impl rustc_session::session::Session>::time
             at ./compiler/rustc_session/src/utils.rs:16:9
  24: rustc_interface::passes::start_codegen
             at ./compiler/rustc_interface/src/passes.rs:1055:19
  25: rustc_interface::queries::Queries::ongoing_codegen::{{closure}}::{{closure}}
             at ./compiler/rustc_interface/src/queries.rs:254:20
  26: rustc_interface::passes::QueryContext::enter::{{closure}}
             at ./compiler/rustc_interface/src/passes.rs:780:42
  27: rustc_middle::ty::context::tls::enter_context::{{closure}}
             at ./compiler/rustc_middle/src/ty/context.rs:1784:50
  28: rustc_middle::ty::context::tls::set_tlv
             at ./compiler/rustc_middle/src/ty/context.rs:1768:9
  29: rustc_middle::ty::context::tls::enter_context
             at ./compiler/rustc_middle/src/ty/context.rs:1784:9
  30: rustc_interface::passes::QueryContext::enter
             at ./compiler/rustc_interface/src/passes.rs:780:9
  31: rustc_interface::queries::Queries::ongoing_codegen::{{closure}}
             at ./compiler/rustc_interface/src/queries.rs:245:13
  32: rustc_interface::queries::Query<T>::compute
             at ./compiler/rustc_interface/src/queries.rs:38:28
  33: rustc_interface::queries::Queries::ongoing_codegen
             at ./compiler/rustc_interface/src/queries.rs:243:9
  34: rustc_driver::run_compiler::{{closure}}::{{closure}}
             at ./compiler/rustc_driver/src/lib.rs:407:13
  35: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
             at ./compiler/rustc_interface/src/queries.rs:394:19
  36: rustc_driver::run_compiler::{{closure}}
             at ./compiler/rustc_driver/src/lib.rs:312:22
  37: rustc_interface::interface::create_compiler_and_run::{{closure}}
             at ./compiler/rustc_interface/src/interface.rs:209:13
  38: rustc_span::with_source_map
             at ./compiler/rustc_span/src/lib.rs:911:5
  39: rustc_interface::interface::create_compiler_and_run
             at ./compiler/rustc_interface/src/interface.rs:203:5
  40: rustc_interface::interface::run_compiler::{{closure}}
             at ./compiler/rustc_interface/src/interface.rs:225:12
  41: rustc_interface::util::setup_callbacks_and_run_in_thread_pool_with_globals::{{closure}}::{{closure}}
             at ./compiler/rustc_interface/src/util.rs:157:13
  42: scoped_tls::ScopedKey<T>::set
             at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/scoped-tls-1.0.0/src/lib.rs:137:9
  43: rustc_span::create_session_globals_then
             at ./compiler/rustc_span/src/lib.rs:105:5
  44: rustc_interface::util::setup_callbacks_and_run_in_thread_pool_with_globals::{{closure}}
             at ./compiler/rustc_interface/src/util.rs:155:9
  45: rustc_interface::util::scoped_thread::{{closure}}
             at ./compiler/rustc_interface/src/util.rs:130:24
```

**Comment by zhassan-aws** [comment]
Neither tests panic with Kani as of `af11d38d93a`, but Kani reports several failures due to unsupported constructs.

**Comment by celinval** [comment]
I accidentally closed this issue when I stated `this does not fix #426` as part of my PR. Re-open.
